### DISCUSSION TITLE
How to target mock-server started with mockoon-cli in powershell and bash to stop it?

### DISCUSSION THREAD

**Discussion by MaciejReszotnik** [mockoon/mockoon#1207] [Help] [answered]
Hi everyone! First time here. Pleased to meet you.

I am considering using Mockoon-cli to set up my mock test server. I installed in with npm i -g mockoon-cli. It runs beautifully. I am able to start it programatically from my code. But I cannot figure out the way to stop the server once it is initiated programatically. 

The standard UI CLI way is to press Ctrl+C. But because of rather convoluted handling of this in Windows and Linux through C# APIs I opted to Close/Kill the processes. The thing is, I cannot find the processes via Get-Processes in powershell. When you run GUI version of Mockoon, "Mockoon" processes are clearly visible. Not so when you run the cli-server.

I anticipate I will face similar problem with bash.

Unfortuantelly due to internal policies I am now allowed to use docker.

I think I am missing something obvious. I would appreciate your help.

Best Regards

**Comment by 255kb** [comment, marked as answer]
Hi, 

Can't you get the pid of the process you start in C#?
If you don't have access to the pid it will be a bit harder as an NPM command line tool is run by Node.js. So, when listing the processes, mockoon-cli will appear as a `node` process.
I am not very familiar with PowerShell or C# development, but I don't think `Get-process` will give you more details than "node", at least this is what I get with multiple Node.js processes running: 

```
PS C:\Users\username> Get-Process -ErrorAction Ignore -Name "node"

Handles  NPM(K)    PM(K)      WS(K)     CPU(s)     Id  SI ProcessName
-------  ------    -----      -----     ------     --  -- -----------
    199      19    57056      36900       0.03    376   1 node
    199      19    56556      36832       0.06   4004   1 node
    199      19    56924      37908       0.20   6980   1 node
    185      17    19748      35744       0.03   7100   1 node
    187     162   380608      47496     378.00   8448   1 node
    445     194   128060      43372       2.77   8580   1 node
    272      64   104900      40940       0.58  25728   1 node
    246      64    54172      40512       0.02  25980   1 node
    362     162  1453784      74424      19.08  27720   1 node
    184      20    63264      38272       0.05  31240   1 node
    199      19    56920      40516       0.14  31588   1 node
    204      69   116392     114148       1.30  31964   1 node <-- ⚠️ this is my mockoon-cli !
```

But I found another command that could allow to filter by the command line and retrieve the processId:

```
PS C:\Users\username> Get-WmiObject Win32_Process -Filter "name = 'node.exe' and commandLine like '%tutorial.json'" | Select-Object -Property CommandLine, ProcessId

CommandLine                                                                                                                                             ProcessId
-----------                                                                                                                                             ---------
...
"C:\Program Files\nodejs\node.exe" C:\Users\gmonn\AppData\Roaming\npm/node_modules/@mockoon/cli/bin/run start -d C:/dev/mockoon/mock-data/tutorial.json     31964
```

**Comment by MaciejReszotnik** [comment]
Thanks for swift reply!

Unfortunately I am unable to identify pId of the process, because I launch the process via powershell. I can identify pId of the powershell console instance, but closing the console does not terminate the server correctly. When I re-launch it I get warning "Error: Port {portFromPreviousRun} is already in use"

I just tried killing off all "node" processes for experiment's sake with process.Kill() and process.Close() exposed in C# Process class. And indeed all of those processes were terminated, and I can set up new servers on old ports without issues. 

Now I need to figure out a way to be more precise with termination.. Unfortunately, while you can get processes by name through Process.GetProcessesByName("node"), the API does not expose any other method to be more precise, and doing this via powershell is a pain in the neck. But that's my problem.

Many thanks for your help :)

I think that as a future improvement mockoon-cli should have mockoon-cli list-servers command to track down all the initialized servers. If the server process was called something different than "node" that would be very helpful as well. I think I will post a feature improvement request later.

**Reply by 255kb** [reply]
There is progress :) 
The CLI was previously using PM2 to manage the processes and we were able to list them with a command. This was removed as there are now many other ways to manage processes. It was also causing a lot of overhead. 
I doubt we will add some mechanism like this back.

But being able to better identify the process makes sense. 
I experimented a bit and it seems that Node.js allows to change the process name with `process.title`. But it doesn't seem to do anything on Windows.
However, the CLI could output its `process.pid`. That could be something we could explore.

If you are running the process with C#, using something like `process.Start()` don't you avec access to a `Process` object containing the `id`?

**Reply by MaciejReszotnik** [reply]
The process object I start is process object of powershell.exe. I only send there cli command to start mockoon-cli server. Here's how it works:

```
public void StartMockServer(int port, string folderPath, string fileName)
  {
      var startInfo = new ProcessStartInfo("powershell.exe")
      {
          UseShellExecute = false,
          WorkingDirectory = Path.Join(ProviderConstants.GetPathToProjectDirectory(), folderPath),
          RedirectStandardInput = true,
          RedirectStandardOutput = true,
          Arguments = $"mockoon-cli start --data  .\\{fileName} --port {port}"
      };
      CliProcess = new Process();
      CliProcess.StartInfo = startInfo;
      CliProcess.Start();
  }
```


Then, when I want to close it I invoke

```
public void StopMockServer()
        {
            if (CliProcess != null)
            {
                CliProcess.StandardInput.Close();
                CliProcess.Close();
            }
        }
```

As you can see, I can easily close the process of the powershell.exe But for some reason it does not terminate the server, even though I think it should. The only thing that terminates server is a piece of code you inspired me to write:

```
public void StopMockServer()
        {
            if (CliProcess != null)
            {
 var processes = Process.GetProcessesByName("node");
                foreach (var process in processes)
                {
                    process.Kill();
                };
                CliProcess.StandardInput.Close();
                CliProcess.Close();
            }
        }
               
```

**Reply by 255kb** [reply]
aybe what you could do instead of starting powershell.exe would be to start node.exe, and pass the location of the CLI script (which is what `mockoon-cli` does behind the scene): 
`C:\Program Files\nodejs\node.exe C:\Users\username\AppData\Roaming\npm\node_modules\@mockoon\cli\bin\run start -d C:/data.json`
